scripts/prices.js

```diff
@@ -1,88 +1,122 @@
- * The generic market data source template, used to build site-specific classes
+ * Oracle's primary instance.
+ *
+ * @todo Allow an array of Oracle instances for better privacy and decentralisation
  */
-export class MarketSource {
-    /** The storage object for raw market data */
-    cData = {};
-
-    /** The name of the market source */
-    strName = '';
-
-    /** The customised API endpoint of the market source */
-    strEndpoint = '';
+export const ORACLE_BASE = 'https://pivxla.bz/oracle/api/v1';
 
+/**
+ * An Oracle instance
+ */
+export class Oracle {
     /**
-     * Ensure a market data cache exists, if not, fetch it and resume
+     * The currencies cache list
+     * @type {Currency[]} Array to store currency objects
      */
-    async ensureCacheExists() {
-        if (!this.cData || !Object.keys(this.cData).length) await this.fetch();
-    }
+    arrCurrencies = [];
 
     /**
-     * Fetches the raw market source data
-     * @returns {Promise<object>}
+     * A lock-like flag which waits until at least once successful "full fetch" of currencies has occurred.
+     * This flag massively lowers bandwidth by only fetching the bulk once, falling to per-currency APIs afterwards.
      */
-    async fetch() {
-        try {
-            return (this.cData = await (await fetch(this.strEndpoint)).json());
-        } catch (e) {
-            console.warn('CoinGecko: Failed to fetch prices!');
-            console.warn(e);
-            return null;
-        }
-    }
-}
-
-/**
- * The CoinGecko market data source
- */
-export class CoinGecko extends MarketSource {
-    constructor() {
-        super();
-        this.strName = 'CoinGecko';
-        this.strEndpoint = COINGECKO_ENDPOINT;
-    }
+    fLoadedCurrencies = false;
 
     /**
-     * Get the price in a specific display currency
-     * @param {string} strCurrency - The CoinGecko-supported display currency
-     * @return {Promise<number>}
+     * Get the price in a specific display currency with extremely low bandwidth
+     * @param {string} strCurrency - The Oracle display currency
+     * @param {boolean} fUseCache - Whether to use local cache or fetch via API
+     * @return {Promise<Currency>}
      */
-    async getPrice(strCurrency) {
-        await this.ensureCacheExists();
-        return this.cData?.market_data?.current_price[strCurrency] || 0;
+    async getPrice(strCurrency, fUseCache = true) {
+        if (fUseCache && this.arrCurrencies.length !== 0) {
+            // We're attempting to read in-cache, and we have some data, let's try finding it
+            return (
+                this.arrCurrencies?.find((a) => a.currency === strCurrency)
+                    ?.value || 0
+            );
+        } else {
+            // Either a refresh request, or no cache available: let's fetch it
+            try {
+                /** @type {Currency} */
+                const cCurrency = await (
+                    await fetch(`${ORACLE_BASE}/price/${strCurrency}`)
```

Comment:
`fetch` doesn't throw an error when the server responds with an error. If the error returns a valid JSON, it may cause an invalid internal state. 
Check that the first fetch call has a true `.ok`.